### PR TITLE
Expand on the set of patterns 'convert for to foreach' detects

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/36305

### CODE REVIEW

File: src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -169,38 +151,71 @@ bool IterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
...
+
+                    // Have to be indexing into the collection.
+                    if (syntaxFacts.AreEquivalent(expr, collectionExpression))
+                        return true;
+                }
+
+                return false;
+            }
+
+            bool IsGoodInvocationExpression(SyntaxNode argumentList)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs:206]: new logic.

**Review by sharwell — APPROVED** [review]

File: src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs [hunk 2]
```diff
@@ -38,9 +38,10 @@ protected override SyntaxList<StatementSyntax> GetBodyStatements(ForStatementSyn
 
         protected override bool TryGetForStatementComponents(
             ForStatementSyntax forStatement,
-            out SyntaxToken iterationVariable, out ExpressionSyntax initializer,
-            out MemberAccessExpressionSyntax memberAccess,
-            out ExpressionSyntax stepValueExpressionOpt,
+            out SyntaxToken iterationVariable,
+            [NotNullWhen(true)] out ExpressionSyntax? initializer,
+            [NotNullWhen(true)] out MemberAccessExpressionSyntax? memberAccess,
+            [NotNullWhen(true)] out ExpressionSyntax? stepValueExpressionOpt,
```
Comment (sharwell) [inline src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs:44]: ❗ This is not a correct annotation.

**Review by sharwell — CHANGES_REQUESTED** [review]
See https://github.com/dotnet/roslyn/pull/66818#discussion_r1103215209

File: src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs]: indeed.  fixed.  not sure how this slipped through either.  @RikkiGibson ?

**Review by sharwell — APPROVED** [review]

File: src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertForToForEach/CSharpConvertForToForEachCodeRefactoringProvider.cs]: It seems like we should flow this through.  'Try' methods deferring to another 'Try' method is a really simple and normal pattern to encounter:-)

**Pull request merged**